pkg/network/ebpf/c/prebuilt/tracer.c

```diff
@@ -1158,52 +1250,300 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
+    //
+    // R0=inv R1=map_value(ks=4,vs=4816) R2=imm5 R3=imm0 R4=imm0 R6=map_value(ks=48,vs=96) R7=imm1 R8=imm0 R9=inv R10=fp
+    ///809: (79) r2 = *(u64 *)(r6 +88)
+    // 810: (7b) *(u64 *)(r0 +88) = r2
+    // R0 invalid mem access 'inv'
+    //
+    // This is because the value range of the R0 register (holding the memory address of the batch) can't be
+    // figured out by the verifier and thus the memory access can't be considered safe during verification time.
+    // I tried different "tricks" to hint the verifier of this range such as:
+    //
+    // * Ensuring 0 <= batch_slot < HTTP_BATCH_SIZE
+    // * Ensuring that &batch <= &batch->txs[batch_state->pos] <= &batch+1
+    // * Setting HTTP_BATCH_SIZEto a power of 2 and doing &batch->txs[batch_slot&(HTTP_BATCH_PAGES-1)]
+    //
+    // Among other things, but nothing really worked on Kernel 4.4
+    // It seems that indeed support for this type of access by the verifier was added later on:
+    // https://patchwork.ozlabs.org/project/netdev/patch/1475074472-23538-1-git-send-email-jbacik@fb.com/
+    //
+    // What is unfortunate about this is not only that enqueing a HTTP transaction is O(HTTP_BATCH_SIZE),
+    // but also that we can't really increase the batch/page size at the moment because that blows up the eBPF *program* size
+#pragma unroll
+    for (int i = 0; i < HTTP_BATCH_SIZE; i++) {
+        if (i == batch_state->pos) {
+            __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
+        }
+    }
+
+    batch_state->pos++;
+    log_debug("http response ended: code: %d duration: %d(ms)\n", http->status_code, (http->response_last_seen-http->request_started)/(1000*1000));
+}
+
+static __always_inline int http_begin_request(http_transaction_t *http, http_state_t new_state, char *buffer) {
+    // This can happen in the context of HTTP keep-alives;
+    if (http->state == HTTP_RESPONDING) {
+        http_end_response(http);
+    }
+
+    log_debug("http request started\n");
+    http->state = new_state;
+    http->request_started = bpf_ktime_get_ns();
+    http->response_last_seen = 0;
+    http->status_code = 0;
+    __builtin_memcpy(&http->request_fragment, buffer, HTTP_BUFFER_SIZE);
+    return 1;
+}
+
+static __always_inline int http_begin_response(http_transaction_t *http, char *buffer) {
+    // We missed the corresponding request so nothing to do
+    if (!(http->state&HTTP_REQUESTING)) {
+        return 0;
+    }
+
+    // Extract the status code from the response fragment
+    // HTTP/1.1 200 OK
+    // _________^^^___
+    // Code below is a bit oddly structured in order to make kernel 4.4 verifier happy
+    __u16 status_code = 0;
+    __u8 space_found = 0;
+#pragma unroll
+    for (int i = 0; i < HTTP_BUFFER_SIZE-1; i++) {
+        if (!space_found && buffer[i] == ' ') {
+            space_found = 1;
+        } else if (space_found && status_code < 100) {
+            status_code = status_code*10 + (buffer[i]-'0');
+        }
```

Comment:
Did you want to `break` once `space_found && status_code >= 100`?

Reply:
Same as above